src/mon/LogMonitor.cc

```diff
@@ -374,6 +379,52 @@ void LogMonitor::log_external_close_fds()
   channel_fds.clear();
 }
 
+/// catch external logs up to summary.version
+void LogMonitor::log_external_backlog()
+{
+  if (!external_log_to) {
+    std::string cur_str;
+    int r = mon.store->read_meta("external_log_to", &cur_str);
+    if (r == 0) {
+      external_log_to = atoll(cur_str.c_str());
+      dout(10) << __func__ << " initialized external_log_to = " << external_log_to
+	       << " (recorded log_to position)" << dendl;
+    } else {
+      // pre-quincy, we assumed that anything through summary.version was
+      // logged externally.
+      external_log_to = summary.version;
+      dout(10) << __func__ << " initialized external_log_to = " << external_log_to
+	       << " (summary v " << summary.version << ")" << dendl;
+    }
+  }
+  assert(external_log_to <= get_last_committed());
+  if (external_log_to == get_last_committed()) {
+    return;
+  }
```

Comment:
i think the purpose of this function is to replay all logs in `(external_log_to, summary.version]`. and leaves the rest of the job, i.e., (summary.version, get_last_committed()]` to its caller. probably, instead of `assert(external_log_to <= get_last_committed());`, we should put
```c++
assert(external_log_to <= summary.version);
```
.

Reply:
hmm, that assert was fine, but the other conditions were a bit off.  the key is that external_log_to is always <= last_committed, but summary.version might be behind that since we only persist it every N expochs.  adjusted the update_from_paxos accordingly so we only log_external() (again) if we haven't already done so while bringing summary.version up to date